### DISCUSSION TITLE
EventState and OnEvent actions relation to eventRef

### DISCUSSION THREAD

**Discussion by globalflea** [serverlessworkflow/specification#524] [Q&A] [answered]
In the example given in the specs, 
```json
{"onEvents": [{
		"eventRefs": ["HighBodyTemperature", "HighBloodPressure"],
		"actions": [{ "functionRef": {
					"refName": "SendTylenolOrder",
					"arguments": {
						"patient": "${ .patientId }"  }}},
			      {	"functionRef": {
					"refName": "CallNurse",
					"arguments": {
						"patient": "${ .patientId }"}}}] }]
}
```

Is there a relationship between items in eventRef and actions? ie. when Exclusive is True, if the event received is "HighBodyTemperature", i call function "SendTylenolOrder",  **but** if the event is "HighBoodPressure" i call function "CallNurse"?  Of if i received **either** event i will call both "SendTylenolOrder" **and** "CallNurse"?  Looking at the example i tend to think there is some relationship

Given that EventState has *N* OnEvents which it must wait for and each OnEvent has *M* EventRefs - does it mean i have to wait for all (*N* * *M*) eventRefs before the EventState is completed? 

Also, if Exclusive is True, and if i received event "HighBodyTemperature", does EventState still continue to wait for "HighBloodPressure"?  i.e. when should i stop waiting for this particular OnEvent or consider this OnEvent completed?

Continuing, if Exclusive is True, and i need to wait for all eventRefs, does it mean that i will call the two functions twice?  i.e. i call  **both** "SendTylenolOrder" and "CallNurse" once when i receive "HighBodyTemperature" and **another time** when i receive "HighBloodPressure"?

**Comment by cdavernas** [comment]
@globalflea The snippet you submitted will trigger both `SendTylenolOrder` and `CallNurse` functions, as you grouped them within the same `trigger`. If you'd like to cherry pick, you'd have to declare two triggers:

```yaml
onEvents:
  - eventsRefs:
    - HighBodyTemperature
    actions:
      - refName: CallNurse
        arguments:
          patient: ${ .patientId }
  - eventsRefs:
    - HighBloodPressure
    actions:
      - refName: SendTylenolOrder
        arguments:
          patient: ${ .patientId }
```

> Is there a relationship between items in eventRef and actions? ie. when Exclusive is True, if the event received is "HighBodyTemperature", i call function "SendTylenolOrder", but if the event is "HighBoodPressure" i call function "CallNurse"? Of if i received either event i will call both "SendTylenolOrder" and "CallNurse"? Looking at the example i tend to think there is some relationship

When `Exclusive` is set to `True`, consuming either even will trigger both actions.

> Given that EventState has N OnEvents which it must wait for and each OnEvent has M EventRefs - does it mean i have to wait for all (N * M) eventRefs before the EventState is completed?

It depends of the value supplied to `Exclusive`. If set to `false`, you'll have to wait for **N * M** events.

> Also, if Exclusive is True, and if i received event "HighBodyTemperature", does EventState still continue to wait for "HighBloodPressure"? i.e. when should i stop waiting for this particular OnEvent or consider this OnEvent completed?

No. Either consumed events should kill the trigger. Other non-consumed events are discarded.

> Continuing, if Exclusive is True, and i need to wait for all eventRefs, does it mean that i will call the two functions twice? i.e. i call both "SendTylenolOrder" and "CallNurse" once when i receive "HighBodyTemperature" and another time when i receive "HighBloodPressure"?

No, in any cases, you are gonna fire the actions of a given trigger once and only once. What `Exclusive` determines, is when my consuming condition is being met (`one of` vs `all of`). If you want to execute an action for each event you are consuming, then you should use multiple non-exclusive triggers.

Hope it answers your question.

**Reply by globalflea** [reply]
Thanks, for explanation. Perhaps the spec can be clearer on this?

**Reply by cdavernas** [reply]
@globalflea You're welcome, any time!
@tsurdilo WDYT?

**Reply by tsurdilo** [reply]
Yeah agree. The reason why event state is like this is because we did not want to go down the route of allowing multiple starting states, like some other DSLs do, and it always ends up in even more confusion and a mess. 
We should raise an issue in spec repo to add more documentation and more samples around this tho, to make it clearer to readers.

**Comment by globalflea** [comment]
To be sure I understand, let say I have an EventState with two OnEvent (OE-A and OE-B) and each of them have two EventRef (so, for OE-A -> OE-A.1 & OE-A.2; OE-B -> OE-B.1 & OE-B.2)

If exclusive is true and the EventState received an event that matches OE-B.1, I conclude that OE-B is completed and the EventState will no longer wait for an event that matches OE-B.2.  However, the EventState will **continue** to wait for events related to OE-A. Is this understanding correct?

If exclusive is false, the EventState will continue to wait for events that will match the remaining 3 events? Ie the EventState will wait till it receives events that matches all OE-A.1, OE-A.2 and OE-B.2.  correct?

**Comment by cdavernas** [comment]
@globalflea For most things, that's my understanding of it, yes. 

For the first case, though, in my implementation the trigger is as well considered implicitly exclusive (the behavior is inherited from the EventState). Is it how it's designed to work, @tsurdilo ?

**Comment by tsurdilo** [comment, marked as answer]
>If exclusive is true and the EventState received an event that matches OE-B.1, I conclude that OE-B is completed and the EventState will no longer wait for an event that matches OE-B.2. However, the EventState will continue to wait for events related to OE-A. Is this understanding correct?

No, the event state does not continue to wait for events related to OE-A. Once one of the eventRefs "fire", regardless of exclusive, the state performs actions and completes. exclusive deals with each individual "eventsRefs" definition, not across them. 

> If exclusive is false, the EventState will continue to wait for events that will match the remaining 3 events? Ie the EventState will wait till it receives events that matches all OE-A.1, OE-A.2 and OE-B.2. correct?

No, if exclusive is false, for OE-A to "fire" it needs A.1 and A.2, for OE-B to fire it needs B1 and B2 events. 
For event sequence, lets say A1, B1, B2, A2, OE-B would get completed ("fire") before Oe-A. So as soon as OE-B fires, its actions are executed, state completes, and workflow continues. 

Hope this makes sense.

**Reply by globalflea** [reply]
would it make more sense to put "exclusive" in each OnEvents rather than at the state level? This mean **each** OnEvent can be either Exclusive or not.  i.e some OnEvent my require that all its EventRefs be matched (OnEvent's exclusive is false) and other OnEvent may require at least one of its EventRef to match (OnEvent's exclusive is true).

Also, "one of" or "all of" seems a lot clearer than "exclusive" and it includes the possibility of a future "n of"?

Finally, perhaps to help with documenting unique fields (eg. Either DataCondition or EventCondition but not both in SwitchState) - would be better to introduce the concept of Enums or some Union type than trying to articulate such relationships with words and notes describing each fields (which will be more prone to typo errors)?

**Reply by tsurdilo** [reply]
1. We can add that. The idea would be to have a default for all (like currently) and each individual could overwrite it. Feel free to open an issue for this to be added.
2. What would be the name of the property? I assume "allOf", "anyOf" would be the values.
3. We do use JSON schema constructs to do this already, so if you define both and do a schema validation it should show as errors. On the spec side we do have to document this with words.
